### PR TITLE
Sheffield | 26-Jan-ITP | Daniel Aderibigbe | Sprint 2 | Coursework/sprint 2

### PR DESCRIPTION
**Pull request by Dan2Clouted**
## Learners, PR Template

Self checklist

- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] My changes meet the requirements of the task
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

## Changelist

Completed all Exercises.

### CODE REVIEW

File: Sprint-2/implement/contains.js [hunk 1]
```diff
@@ -1,3 +1,5 @@
-function contains() {}
+function contains(object, property) {
+  return Object.prototype.hasOwnProperty.call(object, property);
```
Comment (cjyuan) [inline Sprint-2/implement/contains.js:6]: - The function is expected to throw an error or return `false` when object is not an object or is an array.

- Could also explore `Object.hasOwn()`.

File: Sprint-2/implement/contains.test.js [hunk 2]
```diff
@@ -1,35 +1,53 @@
...
+  const obj = { a: 1, b: 2 };
+  expect(contains(obj, "c")).toBe(false);
+});
 
 // Given invalid parameters like an array
 // When passed to contains
 // Then it should return false or throw an error
+test("contains with invalid parameters returns false", () => {
+  expect(contains([], "a")).toBe(false);
+});
```
Comment (cjyuan) [inline Sprint-2/implement/contains.test.js:53]: `expect(contains([], 'a')).toBe(false)` may fail to verify **whether
`contains()` can detect its first argument is an array**.

`contains([], "a")` could also return false because "a" is not a property (or key) of an array.
However, "0", "1", "2" are keys of `[1, 2, 3]`, so it is better to specify the test as
`expect(contains([1, 2, 3], "1")).toBe(false);` (to ensure you are checking what you describe).

You could also add a few more samples to ensure the function can properly handle arguments of different types.

File: Sprint-2/implement/querystring.js [hunk 3]
```diff
@@ -1,13 +1,22 @@
...
-    queryParams[key] = value;
+    const indexOfEquals = pair.indexOf("=");
+
+    if (indexOfEquals === -1) {
+      queryParams[pair] = "";
+    } else {
+      const key = pair.slice(0, indexOfEquals);
+      const value = pair.slice(indexOfEquals + 1);
+      queryParams[key] = value;
+    }
```
Comment (cjyuan) [inline Sprint-2/implement/querystring.js:23]: For each of the following function calls, does your function return the value you expect?
```
parseQueryString("a=b&=&c=d")
parseQueryString("a=")
parseQueryString("=b")
parseQueryString("a=b&&c=d")
```

Note: (You don't have to implement this)
- In real query string, both `key` and `value` are ***percent-encoded*** or ***URL encoded*** in the URL.
For example, the string "5%" is encoded as "5%25". So to get the actual value of "5%25"
(whether it is a key or value in the query string), we need to call a function to decode it.

- You can also explore the `URLSearchParams` API.

File: Sprint-2/implement/tally.js [hunk 4]
```diff
@@ -1,3 +1,19 @@
...
+
+  const result = {};
+
+  for (const item of items) {
+    if (result[item]) {
+      result[item]++;
+    } else {
+      result[item] = 1;
+    }
+  }
```
Comment (cjyuan) [inline Sprint-2/implement/tally.js]: Does the following function call returns the value you expect?
```
tally(["toString", "toString"]);
```

Suggestion: Look up an approach to create an empty object with no inherited properties.

File: Sprint-2/stretch/count-words.js [hunk 5]
```diff
@@ -26,3 +26,24 @@
...
+  for (const word of words) {
+    if (result[word]) {
+      result[word]++;
+    } else {
+      result[word] = 1;
+    }
+  }
+
+  return result;
+}
```
Comment (cjyuan) [inline Sprint-2/stretch/count-words.js:52]: Can you check if your function returns what you expect in the following function calls?

```
countWords("Hello,World! Hello World!");
countWords("constructor constructor");
countWords("          Hello World      ");
```

Note: The spec is not clear about exactly what to expect from these function calls. This is just for self-check.

**Comment by Dan2Clouted** [comment]
Thanks I’ve updated the implementation and tests to better cover the edge cases you pointed out. I handled invalid object inputs for `contains` added extra edge-case checks for `parseQueryString`, used objects without inherited properties for `tally`, and improved `countWords` to cope better with punctuation, repeated spaces, and special property names.

**Comment by cjyuan** [comment]
LGTM.

**Comment by illicitonion** [comment]
Closing PR because the January ITP run has finished. Feel free to re-open if you're still working on it.
